### ISSUE TITLE
Garbage collector on API

### ISSUE THREAD

**Issue by jgsqware** [docker/distribution#1515]
Hello, i've seen in another issues that a garbage collector command exists. 
Is it possible to call it through the api?

My goal is to delete an image, I delete it with the api but the Tags remain in /v2/image/tags/list?

thanks

**Comment by RichardScothern** [comment]
please see https://github.com/docker/distribution/blob/master/docs/gc.md

**Comment by reasonerjt** [comment]
@RichardScothern thanks for the link, but wouldn't it be easier if this feature can be exposed as an API?  Suppose I'm running the registry server in a container and another container as a UI linked to the registry container that user can query the registry to list images.  I can easily integrate this feature on the UI container if the GC is an API, otherwise the UI container needs to talk to docker daemon.

**Comment by jgsqware** [comment]
Totally agreed with @reasonerjt It's exactly the case i'm working on

**Comment by RichardScothern** [comment]
@reasonerjt : the reason for the GC command to not be called from an API is that it requires that the registry not accept any writes while the GC runs (i.e. it is in read-only mode or not running at all).  We anticipate in the future that GC will be live (i.e. it will run periodically while the registry is in use), but before that happens there needs to be some changes made to the way repository metadata is stored.

**Comment by reasonerjt** [comment]
@RichardScothern, thanks for the reply.  
I understand the registry has to be readonly when GC is starts, but I don't think the approach to trigger GC is very related to the required precondition of it...  You can implement GC as an API, and when user calls it, it checks if the registry is read-only mode if not it returns an error code such as 412.

**Comment by RichardScothern** [comment]
I see your point @reasonerjt - though your scenario does not cover multiple registries sharing a backend.  As I mentioned above, GC will become a process that will require no user interaction.  Having it as part of the registry API will mean then removing it from the registry API, which should be as stable as possible.

**Comment by reasonerjt** [comment]
>  "GC will become a process that will require no user interaction."

do you mean it's in the roadmap? Or is there any issue opened to track this work?

**Comment by IevgenKabanets** [comment]
@RichardScothern 
@all 
does GC works OK in current image from Hub (tagged Master)? (https://hub.docker.com/r/distribution/registry/tags)
I run it, nothing is cleaned (while I see some trash to be cleaned).

**Comment by RichardScothern** [comment]
It is coming in release 2.4

You can try the release candidate image: `distribution/registry:2.4.0-rc.1`

**Comment by IevgenKabanets** [comment]
@RichardScothern 
Thanks for proposition. I did try that image, I get loads of DEBUG messages, sort of:

`DEBU[0000] filesystem.List("/docker/registry/v2/blobs/sha256/ff")  environment=development go.version=go1.6 instance.id=f032344c-5c25-4315-962a-56697e933fc0 service=registry trace.duration=18.569µs trace.file=/go/src/github.com/docker/distribution/registry/storage/driver/base/base.go trace.func=github.com/docker/distribution/registry/storage/driver/base.(*Base).List trace.id=1dc6b5c2-2ef5-4320-840f-c85114fe77cb trace.line=150`

But no deletion per se.
I even tried exec into container and run directly there as 
# registry garbage-collect /etc/docker/registry/config.yml

but still the same...

It seems there is not much I can be doing wrong, so it's quite strange situation.

**Comment by RichardScothern** [comment]
A strange situation indeed. @IevgenKabanets .
What was printed at the end of the run

Can you run with the `-d` flag and log level set to `info` and paste the results.

**Comment by IevgenKabanets** [comment]
@RichardScothern thanks for trying to help.
Here goes partial log:
`prpc_ready_build_wbl_oracle
prpc_ready_build_wbl_oracle: marking manifest sha256:62701bc453b0fd33da79610cd7a0835779e1e454c6f6b1bf5b348efbf2c52dee
prpc_ready_build_wbl_oracle: marking blob sha256:4bb326ddb7c4bff16be750e27d6fa46278218a8a0775a6c190835fd02c783886
prpc_ready_build_wbl_oracle: marking configuration sha256:2a0a4d9177e7e5fcb67fc99b2fcc4e5690a7a027221cff08b446f76eb8477c38

398 blobs marked, 0 blobs eligible for deletion`

And so forth for other images. 
We have self-written GC script, that marks quite a few blobs for deletion. So I'm pretty sure there should be cleanup (we store 10 images, aprox. 70 Gb, while now 150 Gb is taken).

**Comment by RichardScothern** [comment]
Do the digests that you deleted through the API appear in that output at all?

**Comment by IevgenKabanets** [comment]
@RichardScothern 
Do you mean digests marked by self-written script?

**Comment by RichardScothern** [comment]
The garbage collector will delete unreferenced data from the storage.  The only way data becomes unreferenced is if you delete the blob or manifest with the delete API.  

So, do the digests you removed with the delete API appear in the output of the garbage collector?

**Comment by IevgenKabanets** [comment]
@RichardScothern 
I think I expected GC to not only collect what was marked for deletion. 
Basically the issue I've tried to address - is existing images getting fatter after few pushes. 
It's less about cleaning up after deletion, more about some leftovers of unknown nature.
Let's say, I have 10 images of 1G. So it should be around 10G of storage taken. It's 20-40. And so on.
Those images are re-created all the time, with diff content, and being pushed into registry (same names and tags, like _registry_url/image_name:latest_).

What I did before to clean is to delete unused BLOBs (simple bash script to check unreferenced tags, corresponding blobs an so on). 
Maybe this is totally wrong, then I'd appreciate the description of correct GC procedure.

**Comment by RichardScothern** [comment]
@IevgenKabanets  the process is described here: https://github.com/docker/distribution/blob/master/docs/gc.md

I understand your scenario:  you consider garbage to be anything that is not referred to by a tag.  The registry considers garbage to be anything that is not referenced by a manifest.  It's possible to push a manifest to the registry without using a tag, and this cannot be considered garbage.

Would the following work in your workflow?
1. HEAD `image:tag` and read the digest from the response
2. push your `image:tag` replacement
3. Use the delete API to remove the digest read in step 1

You could then use the registry garbage collector command.

**Comment by IevgenKabanets** [comment]
@RichardScothern 
Hi Richard, I went that way, looked around APIs once again.
So, here's how it went: 
`curl -I -k https://registry:5000/v2/image_name/manifests/latest` and save Docker-Content-Digest (sha256:blahblah).
`docker push registry:5000/image_name:latest`
`curl -X DELETE -k https://registry:5000/v2/image_name/manifests/sha256:blahblah`

And strangely, last one fails with MANIFEST_UNKNOWN. I've tried on several images. So, size of repo growth, while if to believe APIs - there is nothing to delete. 

Or I'm doing something wrong with calling APIs? 
Thanks!

**Comment by RichardScothern** [comment]
hi @IevgenKabanets : yes, there is something else:  by default the registry will rewrite the manifest to schema1 (which will have a different hash), unless you tell it not to.
https://github.com/docker/distribution/blob/master/docs/spec/api.md#deleting-an-image

TLDR: add this header `Accept: application/vnd.docker.distribution.manifest.v2+json` to your first curl command.

**Comment by IevgenKabanets** [comment]
@RichardScothern so much better now. Thanks, I'll test it out thoroughly now.

**Comment by monikakatiyar16** [comment]
@RichardScothern 
Will docker support running gc via http api or periodically setting the configuration for running gc, using some config file like config.yml? I have read the roadmap describing gc to be implemented for a periodic run and completely automated as opposed to the current implementation. Is there any http api for it included too in any upcoming releases?

**Comment by scottcarey** [comment]
> I understand your scenario: you consider garbage to be anything that is not referred to by a tag. The registry considers garbage to be anything that is not referenced by a manifest. It's possible to push a manifest to the registry without using a tag, and this cannot be considered garbage.

Oh that is downright awful.

Pushing new content to latest is a common practice, especially with continuous automated builds of some sort.  I generate TB's of such cruft a month.  I want gc to clean it up.

>  HEAD image:tag and read the digest from the response
    push your image:tag replacement
    Use the delete API to remove the digest read in step 1

This approach will help in some situations, but what about my 99% full disk today?   

What I need is perhaps a flag to pass to gc that makes it consider anything without a tag garbage.  I have no use in my development / CI environment for anything else.   I don't know what content with a manifest but no tag would be good for.

**Comment by dsaydon90** [comment]
Is this issue still considered as a "question" and not as a "feature request"?
We definitely know that this feature is currently not exist, and we hope to see this feature soon.

**Comment by paullaffitte** [comment]
I hope to see this feature soon too, or at least a "live garbage collector".

To put in context : I run this registry in Kubernetes and I would like to use a CronJob to periodically run garbage collection but it would require the CronJob to delete the pod where the registry is running first and create it back then, which doesn't look very k8s-friendly to me.

**Comment by deleteriousEffect** [comment]
> I hope to see this feature soon too, or at least a "live garbage collector".

@paullaffitte a live garbage collector would most likely require a completely different metadata storage scheme. This isn't off the table as far as I know, but it's a huge undertaking, and it's possible that older images wouldn't be compatible with the new approach.

> To put in context : I run this registry in Kubernetes and I would like to use a CronJob to periodically run garbage collection but it would require the CronJob to delete the pod where the registry is running first and create it back then, which doesn't look very k8s-friendly to me.

You need to put the registry in read-only mode in order to run garbage collect safely, which can be done by flipping a flag in the registry config. I'm not the person to advise on how to admin k8s, but as long as the pods restart to pick up this configuration, you shouldn't need to directly delete the pods. 

> which doesn't look very k8s-friendly to me.

This project is roughly as old as k8s, and there are a number of historical quirks that linger on that didn't anticipate the best practices that would develop as k8s enjoyed its popular success.

**Comment by ArshiAAkhavan** [comment]
Hi
Is there any update in this issue?
I really need a way to automate garbage collection on my registry, API or live GC

**Comment by milosgajdos** [comment]
@ArshiAAkhavan please read what @deleteriousEffect mentioned in her comment.

GC as implemented in this repository is not sufficient enough, certainly not for large registries, and it has also seen very little investment.
